# Customer Churn Prediction & Retention Analytics Platform

## 1. Data Loading

## 2. Initial Data Inspection

## 3. Dataset Overview

## 4. Missing Value Analysis

## 5. Duplicate Analysis

Dataset Source

The Telco Customer Churn dataset contains customer demographics,
subscription details, billing information,
and churn status.

Source:
IBM Sample Telco Customer Churn Dataset

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_excel("../data/raw/Telco_customer_churn.xlsx")

In [3]:
df.head()

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,...,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,...,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,...,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,...,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,...,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices


In [4]:
print(f"Rows: {df.shape[0]}")
print(f"Columns: {df.shape[1]}")

Rows: 7043
Columns: 33


In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 33 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CustomerID         7043 non-null   str    
 1   Count              7043 non-null   int64  
 2   Country            7043 non-null   str    
 3   State              7043 non-null   str    
 4   City               7043 non-null   str    
 5   Zip Code           7043 non-null   int64  
 6   Lat Long           7043 non-null   str    
 7   Latitude           7043 non-null   float64
 8   Longitude          7043 non-null   float64
 9   Gender             7043 non-null   str    
 10  Senior Citizen     7043 non-null   str    
 11  Partner            7043 non-null   str    
 12  Dependents         7043 non-null   str    
 13  Tenure Months      7043 non-null   int64  
 14  Phone Service      7043 non-null   str    
 15  Multiple Lines     7043 non-null   str    
 16  Internet Service   7043 non-null   

In [6]:
df.describe(include="all").T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
CustomerID,7043,7043,3668-QPYBK,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Count,7043.0,NaN,NaN,NaN,1.0,0.0,1.0,1.0,1.0,1.0,1.0
Country,7043,1,United States,7043,NaN,NaN,NaN,NaN,NaN,NaN,NaN
State,7043,1,California,7043,NaN,NaN,NaN,NaN,NaN,NaN,NaN
City,7043,1129,Los Angeles,305,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Zip Code,7043.0,NaN,NaN,NaN,93521.964646,1865.794555,90001.0,92102.0,93552.0,95351.0,96161.0
Lat Long,7043,1652,"33.964131, -118.272783",5,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Latitude,7043.0,NaN,NaN,NaN,36.282441,2.455723,32.555828,34.030915,36.391777,38.224869,41.962127
Longitude,7043.0,NaN,NaN,NaN,-119.79888,2.157889,-124.301372,-121.815412,-119.730885,-118.043237,-114.192901
Gender,7043,2,Male,3555,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [7]:
missing = (
    df.isnull()
      .sum()
      .to_frame(name="Missing Values")
)

missing["Percentage"] = (
    missing["Missing Values"] / len(df) * 100
).round(2)

missing.sort_values(
    by="Missing Values",
    ascending=False
)

,Missing Values,Percentage
Churn Reason,5174,73.46
CustomerID,0,0.00
Count,0,0.00
State,0,0.00
Country,0,0.00
Zip Code,0,0.00
Lat Long,0,0.00
Latitude,0,0.00
City,0,0.00
Gender,0,0.00


In [8]:
print(f"Duplicate Rows: {df.duplicated().sum()}")

Duplicate Rows: 0


# Data Audit Summary

### Dataset Size
- Records: 7,043
- Features: 33

### Data Quality Assessment
- Duplicate Records: 0
- Missing Values: Present only in `Churn Reason` (73.46%).
- Missing values are expected because only churned customers have a recorded reason for leaving.

### Initial Observations
- Dataset is highly complete.
- Contains demographic, geographic, service, billing, churn, and customer lifetime value information.
- Suitable for predictive analytics and business intelligence.

## Target Variable

Target:

Churn Value

0 → Customer Retained

1 → Customer Churned

The objective of this project is to predict
whether a customer is likely to churn.

## Feature Selection Strategy

To prevent data leakage and improve model generalization, features were categorized into:

- Machine Learning Features
- Business Analysis Features
- Identifier / Redundant Features
- Data Leakage Features

Only information that would realistically be available before a customer churns will be used for predictive modeling.

In [9]:
feature_matrix = pd.DataFrame({
    "Column": df.columns
})

feature_matrix

,Column
0,CustomerID
1,Count
2,Country
3,State
4,City
5,Zip Code
6,Lat Long
7,Latitude
8,Longitude
9,Gender


In [10]:
feature_matrix = pd.DataFrame({
    "Column": [
        "CustomerID", "Count", "Country", "State", "City", "Zip Code",
        "Lat Long", "Latitude", "Longitude", "Gender", "Senior Citizen",
        "Partner", "Dependents", "Tenure Months", "Phone Service",
        "Multiple Lines", "Internet Service", "Online Security",
        "Online Backup", "Device Protection", "Tech Support",
        "Streaming TV", "Streaming Movies", "Contract",
        "Paperless Billing", "Payment Method", "Monthly Charges",
        "Total Charges", "Churn Label", "Churn Value",
        "Churn Score", "CLTV", "Churn Reason"
    ],

    "Category": [
        "Identifier","Constant","Constant","Geographic","Geographic","Geographic",
        "Geographic","Geographic","Geographic","Demographic","Demographic",
        "Demographic","Demographic","Customer History","Service",
        "Service","Service","Service","Service","Service",
        "Service","Service","Service","Subscription",
        "Billing","Billing","Financial",
        "Financial","Target","Target",
        "Leakage","Business","Leakage"
    ],

    "ML Usage": [
        "Drop","Drop","Drop","No","No","No",
        "Drop","No","No","Yes","Yes",
        "Yes","Yes","Yes","Yes",
        "Yes","Yes","Yes","Yes","Yes",
        "Yes","Yes","Yes","Yes",
        "Yes","Yes","Yes",
        "Yes","No","Target",
        "No","No","No"
    ],

    "Dashboard Usage": [
        "No","No","No","Yes","Yes","Yes",
        "No","Yes","Yes","Yes","Yes",
        "Yes","Yes","Yes","Yes",
        "Yes","Yes","Yes","Yes","Yes",
        "Yes","Yes","Yes","Yes",
        "Yes","Yes","Yes",
        "Yes","Yes","No",
        "Yes","Yes","Yes"
    ],

    "Decision": [
        "Drop","Drop","Drop","Dashboard Only","Dashboard Only","Dashboard Only",
        "Drop","Dashboard Only","Dashboard Only","Keep","Keep",
        "Keep","Keep","Keep","Keep",
        "Keep","Keep","Keep","Keep","Keep",
        "Keep","Keep","Keep","Keep",
        "Keep","Keep","Keep",
        "Keep","Dashboard Only","Target Variable",
        "Dashboard Only","Dashboard Only","Dashboard Only"
    ]
})

feature_matrix

,Column,Category,ML Usage,Dashboard Usage,Decision
0,CustomerID,Identifier,Drop,No,Drop
1,Count,Constant,Drop,No,Drop
2,Country,Constant,Drop,No,Drop
3,State,Geographic,No,Yes,Dashboard Only
4,City,Geographic,No,Yes,Dashboard Only
5,Zip Code,Geographic,No,Yes,Dashboard Only
6,Lat Long,Geographic,Drop,No,Drop
7,Latitude,Geographic,No,Yes,Dashboard Only
8,Longitude,Geographic,No,Yes,Dashboard Only
9,Gender,Demographic,Yes,Yes,Keep


In [11]:
df.columns.tolist()

['CustomerID',
 'Count',
 'Country',
 'State',
 'City',
 'Zip Code',
 'Lat Long',
 'Latitude',
 'Longitude',
 'Gender',
 'Senior Citizen',
 'Partner',
 'Dependents',
 'Tenure Months',
 'Phone Service',
 'Multiple Lines',
 'Internet Service',
 'Online Security',
 'Online Backup',
 'Device Protection',
 'Tech Support',
 'Streaming TV',
 'Streaming Movies',
 'Contract',
 'Paperless Billing',
 'Payment Method',
 'Monthly Charges',
 'Total Charges',
 'Churn Label',
 'Churn Value',
 'Churn Score',
 'CLTV',
 'Churn Reason']

In [12]:
df.dtypes

CustomerID               str
Count                  int64
Country                  str
State                    str
City                     str
Zip Code               int64
Lat Long                 str
Latitude             float64
Longitude            float64
Gender                   str
Senior Citizen           str
Partner                  str
Dependents               str
Tenure Months          int64
Phone Service            str
Multiple Lines           str
Internet Service         str
Online Security          str
Online Backup            str
Device Protection        str
Tech Support             str
Streaming TV             str
Streaming Movies         str
Contract                 str
Paperless Billing        str
Payment Method           str
Monthly Charges      float64
Total Charges         object
Churn Label              str
Churn Value            int64
Churn Score            int64
CLTV                   int64
Churn Reason             str
dtype: object

In [15]:
for col in df.select_dtypes(include="object"):
    if col != "CustomerID":
        print("\n" + "=" * 50)
        print(col)
        print(df[col].unique())


Country
<StringArray>
['United States']
Length: 1, dtype: str

State
<StringArray>
['California']
Length: 1, dtype: str

City
<StringArray>
[    'Los Angeles',   'Beverly Hills', 'Huntington Park',         'Lynwood',
  'Marina Del Rey',       'Inglewood',    'Santa Monica',        'Torrance',
        'Whittier',        'La Habra',
 ...
      'Janesville',      'Litchfield',        'Loyalton',        'Madeline',
    'Markleeville',         'Milford',         'Calpine',        'Standish',
        'Tulelake',  'Olympic Valley']
Length: 1129, dtype: str

Lat Long
<StringArray>
['33.964131, -118.272783',  '34.059281, -118.30742', '34.048013, -118.293953',
 '34.062125, -118.315709', '34.039224, -118.266293', '34.066367, -118.309868',
  '34.02381, -118.156582', '34.066303, -118.435479', '34.099869, -118.326843',
 '34.089953, -118.294824',
 ...
 '40.294034, -120.512622', '40.507272, -120.338228', '39.637471, -120.226338',
 '41.042003, -120.506086',  '38.735789, -119.85798', '40.181279, -120.3

C:\Users\Tanvi\AppData\Local\Temp\ipykernel_13264\3248591169.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  for col in df.select_dtypes(include="object"):


In [14]:
constant_columns = [
    col for col in df.columns
    if df[col].nunique() == 1
]

constant_columns

['Count', 'Country', 'State']

In [16]:
clean_df = df.copy()

In [17]:
constant_columns = ["Count", "Country", "State"]

clean_df.drop(columns=constant_columns, inplace=True)

print(clean_df.shape)

(7043, 30)


In [18]:
clean_df.columns

Index(['CustomerID', 'City', 'Zip Code', 'Lat Long', 'Latitude', 'Longitude',
       'Gender', 'Senior Citizen', 'Partner', 'Dependents', 'Tenure Months',
       'Phone Service', 'Multiple Lines', 'Internet Service',
       'Online Security', 'Online Backup', 'Device Protection', 'Tech Support',
       'Streaming TV', 'Streaming Movies', 'Contract', 'Paperless Billing',
       'Payment Method', 'Monthly Charges', 'Total Charges', 'Churn Label',
       'Churn Value', 'Churn Score', 'CLTV', 'Churn Reason'],
      dtype='str')

# Data Cleaning Log

### Cleaning Step 1
Removed constant-value columns:

- Count
- Country
- State

Reason:
These columns contain only one unique value across all customer records and therefore provide no analytical or predictive value.

In [19]:
clean_df.to_csv(
    "../data/cleaned/Telco_customer_churn_clean.csv",
    index=False
)

print("Clean dataset saved successfully!")

Clean dataset saved successfully!
